mypyc/genops.py

```diff
@@ -1702,33 +1710,60 @@ def assign(self, target: Union[Register, AssignmentTarget],
         else:
             assert False, 'Unsupported assignment target'
 
+    def process_iterator_tuple_assignment_helper(self,
+                                                 litem: AssignmentTarget,
+                                                 ritem: Value, line: int) -> None:
+        error_block, ok_block = BasicBlock(), BasicBlock()
+        self.add(Branch(ritem, error_block, ok_block, Branch.IS_ERROR))
+
+        self.activate_block(error_block)
+        self.add(RaiseStandardError(RaiseStandardError.VALUE_ERROR,
+                                    'not enough values to unpack', line))
+        self.add(Unreachable())
+
+        self.activate_block(ok_block)
+        self.assign(litem, ritem, line)
+
     def process_iterator_tuple_assignment(self,
                                           target: AssignmentTargetTuple,
                                           rvalue_reg: Value,
                                           line: int) -> None:
+
         iterator = self.primitive_op(iter_op, [rvalue_reg], line)
-        for litem in target.items:
-            error_block, ok_block = BasicBlock(), BasicBlock()
+
+        # This may be the whole lvalue list if there is no starred value
+        prestar_vals = list(itertools.takewhile(lambda val: val != target.starred, target.items))
+        rest = target.items[len(prestar_vals):]
+
+        # Assign values before the first starred value
+        for litem in prestar_vals:
             ritem = self.primitive_op(next_op, [iterator], line)
-            self.add(Branch(ritem, error_block, ok_block, Branch.IS_ERROR))
+            self.process_iterator_tuple_assignment_helper(litem, ritem, line)
+
+        # Assign the starred value and all values after it
+        if rest:
+            iter_list = self.primitive_op(to_list, [iterator], line)
+            starred = rest[0]
+            for litem in reversed(rest[1:]):
```

Comment:
Thinking about this a bit more, I think we should do an extra check here: if there aren't enough elements and `pop` fails, that will raise `IndexError`. But not enough values to unpack is a `ValueError` in python. mypyc definitely does not always produce the same exception messages but I think we ought to at least try to produce the same exceptions.

So I think we should do a length check to make sure there are enough elements and raise ValueError if there aren't.